File: 1300-1399/1301. Number of Paths with Maximum Score/Solution.cpp
// Approach-1 (Bottom-Up Dynamic Programming)
// T.C : O(n²)
// S.C : O(n²)

class Solution
{
public:
  vector<int> pathsWithMaxScore(vector<string> &board)
  {

    int n = board.size();

    const int MOD = 1000000007;

    vector<vector<int>> maxScore(
        n,
        vector<int>(n, -1));

    vector<vector<int>> pathCount(
        n,
        vector<int>(n, 0));

    maxScore[n - 1][n - 1] = 0;

    pathCount[n - 1][n - 1] = 1;

    for (int row = n - 1; row >= 0; row--)
    {

      for (int col = n - 1; col >= 0; col--)
      {

        if (board[row][col] == 'X')
        {

          continue;
        }

        if (
            row == n - 1 &&
            col == n - 1)
        {

          continue;
        }

        int bestScore = -1;

        int ways = 0;

        vector<pair<int, int>> directions = {
            {row + 1, col},
            {row, col + 1},
            {row + 1, col + 1}};

        for (auto &next : directions)
        {

          int newRow = next.first;

          int newCol = next.second;

          if (
              newRow >= n ||
              newCol >= n ||
              maxScore[newRow][newCol] == -1)
          {

            continue;
          }

          if (
              maxScore[newRow][newCol] >
              bestScore)
          {

            bestScore =
                maxScore[newRow][newCol];

            ways =
                pathCount[newRow][newCol];
          }
          else if (
              maxScore[newRow][newCol] ==
              bestScore)
          {

            ways =
                (ways +
                 pathCount[newRow][newCol]) %
                MOD;
          }
        }

        if (bestScore == -1)
        {

          continue;
        }

        if (
            isdigit(board[row][col]))
        {

          bestScore +=
              board[row][col] - '0';
        }

        maxScore[row][col] =
            bestScore;

        pathCount[row][col] =
            ways;
      }
    }

    if (pathCount[0][0] == 0)
    {

      return {0, 0};
    }

    return {
        maxScore[0][0],
        pathCount[0][0]};
  }
};

/*
//Approach-2 (Top-Down Dynamic Programming + Memoization)
//T.C : O(n²)
//S.C : O(n²)

class Solution {
public:

    const int MOD = 1000000007;

    vector<vector<int>> score;

    vector<vector<int>> ways;

    pair<int,int> dfs(
        int row,
        int col,
        vector<string>& board
    ) {

        int n = board.size();

        if (
            row < 0 ||
            col < 0 ||
            board[row][col] == 'X'
        ) {

            return {-1, 0};
        }

        if (
            row == n - 1 &&
            col == n - 1
        ) {

            return {0, 1};
        }

        if (score[row][col] != -2) {

            return {
                score[row][col],
                ways[row][col]
            };
        }

        vector<pair<int,int>> moves = {
            {row + 1, col},
            {row, col + 1},
            {row + 1, col + 1}
        };

        int bestScore = -1;

        long long totalWays = 0;

        for (auto &next : moves) {

            auto result =
                dfs(
                    next.first,
                    next.second,
                    board
                );

            if (result.first == -1) {

                continue;
            }

            if (result.first > bestScore) {

                bestScore = result.first;

                totalWays = result.second;

            }
            else if (
                result.first == bestScore
            ) {

                totalWays =
                    (totalWays +
                     result.second)
                    % MOD;
            }
        }

        if (bestScore == -1) {

            score[row][col] = -1;

            ways[row][col] = 0;

            return {-1, 0};
        }

        if (
            isdigit(board[row][col])
        ) {

            bestScore +=
                board[row][col] - '0';
        }

        score[row][col] = bestScore;

        ways[row][col] = totalWays;

        return {
            bestScore,
            (int) totalWays
        };
    }

    vector<int> pathsWithMaxScore(
        vector<string>& board
    ) {

        int n = board.size();

        score.assign(
            n,
            vector<int>(n, -2)
        );

        ways.assign(
            n,
            vector<int>(n, 0)
        );

        auto answer =
            dfs(
                0,
                0,
                board
            );

        if (answer.second == 0) {

            return {0, 0};
        }

        return {
            answer.first,
            answer.second
        };
    }
};
*/

File: 1300-1399/1301. Number of Paths with Maximum Score/Solution.java
//Approach-1 (Bottom-Up Dynamic Programming)
//T.C : O(n²)
//S.C : O(n²)

import java.util.*;

class Solution {

  public int[] pathsWithMaxScore(List<String> board) {

    int size = board.size();

    int mod = 1000000007;

    int[][] maxScore = new int[size][size];

    int[][] pathCount = new int[size][size];

    for (int i = 0; i < size; i++) {

      Arrays.fill(maxScore[i], -1);
    }

    maxScore[size - 1][size - 1] = 0;

    pathCount[size - 1][size - 1] = 1;

    for (int row = size - 1; row >= 0; row--) {

      for (int col = size - 1; col >= 0; col--) {

        if (board.get(row).charAt(col) == 'X') {

          continue;
        }

        if (row == size - 1 &&
            col == size - 1) {

          continue;
        }

        int bestScore = -1;

        int ways = 0;

        if (row + 1 < size &&
            maxScore[row + 1][col] != -1) {

          if (maxScore[row + 1][col] > bestScore) {

            bestScore = maxScore[row + 1][col];

            ways = pathCount[row + 1][col];

          } else if (maxScore[row + 1][col] == bestScore) {

            ways = (ways +
                pathCount[row + 1][col])
                % mod;
          }
        }

        if (col + 1 < size &&
            maxScore[row][col + 1] != -1) {

          if (maxScore[row][col + 1] > bestScore) {

            bestScore = maxScore[row][col + 1];

            ways = pathCount[row][col + 1];

          } else if (maxScore[row][col + 1] == bestScore) {

            ways = (ways +
                pathCount[row][col + 1])
                % mod;
          }
        }

        if (row + 1 < size &&
            col + 1 < size &&
            maxScore[row + 1][col + 1] != -1) {

          if (maxScore[row + 1][col + 1] > bestScore) {

            bestScore = maxScore[row + 1][col + 1];

            ways = pathCount[row + 1][col + 1];

          } else if (maxScore[row + 1][col + 1] == bestScore) {

            ways = (ways +
                pathCount[row + 1][col + 1])
                % mod;
          }
        }

        if (bestScore == -1) {

          continue;
        }

        char current = board.get(row).charAt(col);

        if (Character.isDigit(current)) {

          bestScore += current - '0';
        }

        maxScore[row][col] = bestScore;

        pathCount[row][col] = ways;
      }
    }

    if (pathCount[0][0] == 0) {

      return new int[] { 0, 0 };
    }

    return new int[] {
        maxScore[0][0],
        pathCount[0][0]
    };
  }
}

/*
 * //Approach-2 (Top-Down Dynamic Programming + Memoization)
 * //T.C : O(n²)
 * //S.C : O(n²)
 * 
 * import java.util.*;
 * 
 * class Solution {
 * 
 * private final int MOD = 1000000007;
 * 
 * private int[][] score;
 * 
 * private int[][] ways;
 * 
 * private boolean[][] visited;
 * 
 * public int[] pathsWithMaxScore(List<String> board) {
 * 
 * int n = board.size();
 * 
 * score = new int[n][n];
 * 
 * ways = new int[n][n];
 * 
 * visited = new boolean[n][n];
 * 
 * int[] answer = dfs(
 * 0,
 * 0,
 * board
 * );
 * 
 * if (answer[1] == 0) {
 * 
 * return new int[]{0, 0};
 * }
 * 
 * return answer;
 * }
 * 
 * private int[] dfs(
 * int row,
 * int col,
 * List<String> board
 * ) {
 * 
 * int n = board.size();
 * 
 * if (
 * row >= n ||
 * col >= n ||
 * board.get(row).charAt(col) == 'X'
 * ) {
 * 
 * return new int[]{-1, 0};
 * }
 * 
 * if (
 * row == n - 1 &&
 * col == n - 1
 * ) {
 * 
 * return new int[]{0, 1};
 * }
 * 
 * if (visited[row][col]) {
 * 
 * return new int[]{
 * score[row][col],
 * ways[row][col]
 * };
 * }
 * 
 * visited[row][col] = true;
 * 
 * int bestScore = -1;
 * 
 * int totalWays = 0;
 * 
 * int[][] directions = {
 * {1, 0},
 * {0, 1},
 * {1, 1}
 * };
 * 
 * for (int[] direction : directions) {
 * 
 * int[] next =
 * dfs(
 * row + direction[0],
 * col + direction[1],
 * board
 * );
 * 
 * if (next[0] == -1) {
 * 
 * continue;
 * }
 * 
 * if (next[0] > bestScore) {
 * 
 * bestScore = next[0];
 * 
 * totalWays = next[1];
 * 
 * } else if (
 * next[0] == bestScore
 * ) {
 * 
 * totalWays =
 * (totalWays + next[1])
 * % MOD;
 * }
 * }
 * 
 * if (bestScore == -1) {
 * 
 * score[row][col] = -1;
 * 
 * ways[row][col] = 0;
 * 
 * return new int[]{-1, 0};
 * }
 * 
 * char current =
 * board.get(row).charAt(col);
 * 
 * if (
 * Character.isDigit(current)
 * ) {
 * 
 * bestScore +=
 * current - '0';
 * }
 * 
 * score[row][col] = bestScore;
 * 
 * ways[row][col] = totalWays;
 * 
 * return new int[]{
 * bestScore,
 * totalWays
 * };
 * }
 * }
 */